File: app/test/domain/streak_calculator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/domain/streak_calculator.dart';

void main() {
  group('evaluateStreakOnActive', () {
    test('first ever active day sets streak to 1', () {
      final r = evaluateStreakOnActive(
        currentStreakDays: 0,
        longestStreakDays: 0,
        freezesRemaining: 2,
        lastActiveDate: null,
        now: DateTime(2026, 8, 14),
      );
      expect(r.newCurrentDays, 1);
      expect(r.streakBroken, isFalse);
      expect(r.freezesConsumed, 0);
    });

    test('consecutive day increments streak', () {
      final r = evaluateStreakOnActive(
        currentStreakDays: 5,
        longestStreakDays: 10,
        freezesRemaining: 2,
        lastActiveDate: DateTime(2026, 8, 13),
        now: DateTime(2026, 8, 14),
      );
      expect(r.newCurrentDays, 6);
      expect(r.streakBroken, isFalse);
    });

    test('missed 1 day consumes freeze and preserves streak', () {
      final r = evaluateStreakOnActive(
        currentStreakDays: 5,
        longestStreakDays: 10,
        freezesRemaining: 2,
        lastActiveDate: DateTime(2026, 8, 12),
        now: DateTime(2026, 8, 14),
      );
      expect(r.newCurrentDays, 6);
      expect(r.freezesConsumed, 1);
      expect(r.streakBroken, isFalse);
    });

    test('missed 2 days with no freezes resets streak', () {
      final r = evaluateStreakOnActive(
        currentStreakDays: 5,
        longestStreakDays: 10,
        freezesRemaining: 0,
        lastActiveDate: DateTime(2026, 8, 11),
        now: DateTime(2026, 8, 14),
      );
      expect(r.newCurrentDays, 1);
      expect(r.streakBroken, isTrue);
      expect(r.newLongestDays, 10); // 不变
    });

    test('same-day active does nothing', () {
      final r = evaluateStreakOnActive(
        currentStreakDays: 5,
        longestStreakDays: 10,
        freezesRemaining: 2,
        lastActiveDate: DateTime(2026, 8, 14),
        now: DateTime(2026, 8, 14),
      );
      expect(r.newCurrentDays, 5);
      expect(r.freezesConsumed, 0);
    });
  });

  group('shouldAwardFreeze', () {
    test('awards when 30 days passed since last award', () {
      final result = shouldAwardFreeze(
        lastFreezeAwardDate: DateTime(2026, 7, 15),
        now: DateTime(2026, 8, 14),
      );
      expect(result, isTrue);
    });

    test('does not award within 30 days', () {
      final result = shouldAwardFreeze(
        lastFreezeAwardDate: DateTime(2026, 8, 1),
        now: DateTime(2026, 8, 14),
      );
      expect(result, isFalse);
    });

    test('awards when never awarded and 30 days passed', () {
      final result = shouldAwardFreeze(
        lastFreezeAwardDate: null,
        now: DateTime(2026, 8, 14),
      );
      expect(result, isFalse); // 首次安装不立即给
    });
  });
}

File: app/android/app/src/test/java/com/gymcoach/gym_coach/domain/StreakCalculatorTest.kt
package com.gymcoach.gym_coach.domain

import com.gymcoach.gym_coach.domain.streak.StreakCalculator
import com.google.common.truth.Truth.assertThat
import java.time.LocalDate
import org.junit.Test

class StreakCalculatorTest {

    private val today = LocalDate.of(2026, 8, 15)
    private val yesterday = today.minusDays(1)
    private val twoDaysAgo = today.minusDays(2)

    @Test
    fun `first active day sets current to 1 and freezes to 0`() {
        val r = StreakCalculator.evaluateStreakOnActive(
            currentStreakDays = 0,
            longestStreakDays = 0,
            freezesRemaining = 2,
            lastActiveDate = null,
            now = today
        )
        assertThat(r.newCurrentDays).isEqualTo(1)
        assertThat(r.newLongestDays).isEqualTo(1)
        assertThat(r.freezesConsumed).isEqualTo(0)
        assertThat(r.streakBroken).isFalse()
    }

    @Test
    fun `consecutive day increments and never breaks`() {
        val r = StreakCalculator.evaluateStreakOnActive(
            currentStreakDays = 3,
            longestStreakDays = 5,
            freezesRemaining = 2,
            lastActiveDate = yesterday,
            now = today
        )
        assertThat(r.newCurrentDays).isEqualTo(4)
        assertThat(r.newLongestDays).isEqualTo(5)
        assertThat(r.freezesConsumed).isEqualTo(0)
        assertThat(r.streakBroken).isFalse()
    }

    @Test
    fun `skip one day consumes one freeze`() {
        val r = StreakCalculator.evaluateStreakOnActive(
            currentStreakDays = 3,
            longestStreakDays = 3,
            freezesRemaining = 2,
            lastActiveDate = twoDaysAgo,
            now = today
        )
        assertThat(r.newCurrentDays).isEqualTo(4)
        assertThat(r.freezesConsumed).isEqualTo(1)
        assertThat(r.streakBroken).isFalse()
    }

    @Test
    fun `skip two days with no freezes resets and marks broken`() {
        val r = StreakCalculator.evaluateStreakOnActive(
            currentStreakDays = 5,
            longestStreakDays = 7,
            freezesRemaining = 0,
            lastActiveDate = today.minusDays(3),
            now = today
        )
        assertThat(r.newCurrentDays).isEqualTo(1)
        assertThat(r.newLongestDays).isEqualTo(7)
        assertThat(r.freezesConsumed).isEqualTo(0)
        assertThat(r.streakBroken).isTrue()
    }

    @Test
    fun `same day is a no-op`() {
        val r = StreakCalculator.evaluateStreakOnActive(
            currentStreakDays = 5,
            longestStreakDays = 5,
            freezesRemaining = 2,
            lastActiveDate = today,
            now = today
        )
        assertThat(r.newCurrentDays).isEqualTo(5)
        assertThat(r.newLongestDays).isEqualTo(5)
        assertThat(r.freezesConsumed).isEqualTo(0)
    }

    @Test
    fun `shouldAwardFreeze only after 30 days`() {
        val awarded = today.minusDays(45)
        val recent = today.minusDays(10)
        assertThat(StreakCalculator.shouldAwardFreeze(awarded, today)).isTrue()
        assertThat(StreakCalculator.shouldAwardFreeze(recent, today)).isFalse()
        assertThat(StreakCalculator.shouldAwardFreeze(null, today)).isFalse()
    }
}
